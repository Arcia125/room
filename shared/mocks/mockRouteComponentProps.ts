/**
 * @description Mock props for Children of Route components.
 */
import { RouteComponentProps, match } from 'react-router';
import * as history from 'history';

const mockLocation: history.Location = {
  hash: '',
  pathname: '/',
  search: '/',
  state: {},
};

const mockHistory: history.History = {
  push: jest.fn(),
  length: 1,
  block: jest.fn(),
  action: 'PUSH',
  createHref: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  listen: jest.fn(),
  location: mockLocation,
  replace: jest.fn(),
};

const mockMatch: match = {
  isExact: true,
  params: {},
  path: '/',
  url: '/',
};

/**
 * @description mocked version of props passed from react router Routes to their children
 * @example
 * render(<MyComponent {...mockRouteComponentProps} />)
 */
export const mockRouteComponentProps: RouteComponentProps = {
  history: mockHistory,
  location: mockLocation,
  match: mockMatch,
};
