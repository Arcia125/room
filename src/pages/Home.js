import React from 'react';
import PropTypes from 'prop-types';
import { useMutation } from '@apollo/react-hooks';

import logo from '../logo.svg';
import { ADD_ROOM } from '../graphql/addRoom';

const roomNames = [
  'test1',
  'test2',
];

const Home = ({ history }) => {
  const [addRoom, {data, loading, error }] = useMutation(ADD_ROOM, { variables: { name: roomNames[Math.random() > .5 ? 0 : 1] }});

  React.useEffect(() => {
    if (data && data.addRoom && data.addRoom.id) {
      history.push(`/r/${data.addRoom.id}`);
    }
  }, [data && data.addRoom && data.addRoom.id]);

  const handleCreateRoom = () => addRoom();

  return (
    <div className="home-page">
      <header className="home-page-header">
        <img src={logo} className="home-page-logo" alt="logo" />
        <p>Create a room to get started</p>
        <button className="create-room-button" onClick={handleCreateRoom}>create a room</button>
        <pre>{JSON.stringify(data, null, 2)}</pre>
      </header>
    </div>
  );
};

Home.propTypes = {};

export default Home;
