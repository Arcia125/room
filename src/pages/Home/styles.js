import styled from 'styled-components';

const HomePage = styled.div`
  text-align: center;

  .home-page-logo {
    animation: home-page-logo-spin infinite 20s linear;
    height: 40vmin;
    pointer-events: none;
  }

  .home-page-header {
    background-color: #282c34;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: white;
  }

  @keyframes home-page-logo-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
`;

export { HomePage };
