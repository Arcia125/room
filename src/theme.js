const theme = {
  spacing: {
    0: '0px',
    1: '8px',
    2: '16px',
    3: '32px'
  }
};

export { theme };
