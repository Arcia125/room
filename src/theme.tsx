import React, { ReactElement } from 'react';
import { ThemeProvider } from 'styled-components';

const theme = {
  spacing: {
    0: '0px',
    1: '8px',
    2: '16px',
    3: '32px',
  },
  colors: {
    gray: '#B3B2C0',
    grayDark: '#1f1f1f',
    lightBlue: '#AEAFE8',
    blue: '#AC2FC9',
    lightPurple: '#EFEEF8',
    purple: '#9056D4',
    pink: '#DD6589',
    bleh: '#8A4966',
  },
};

const Provider: React.FunctionComponent<{ children: ReactElement }> = ({
  children,
  ...restProps
}) => (
  <ThemeProvider theme={theme} {...restProps}>
    {children}
  </ThemeProvider>
);

export { theme, Provider };
