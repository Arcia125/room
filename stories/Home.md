# Home

<Giphy query="landing" />

The landing page of the site.
