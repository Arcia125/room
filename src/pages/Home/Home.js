import React from 'react';
import PropTypes from 'prop-types';
import { useMutation } from '@apollo/react-hooks';

import logo from '../../logo.svg';
import { ADD_ROOM } from '../../graphql/addRoom';
import { HomePage } from './styles';
import { Button } from '../../components/styles/Button';

const roomNames = ['test1', 'test2'];

const Home = ({ history }) => {
  const [addRoom, addRoomMutation] = useMutation(ADD_ROOM, {
    variables: { name: roomNames[Math.random() > 0.5 ? 0 : 1] }
  });

  const addedRoomId = addRoomMutation.data && addRoomMutation.data.addRoom.id;

  React.useEffect(() => {
    if (addedRoomId || addedRoomId === 0) {
      history.push(`/r/${addedRoomId}`);
    }
  }, [addedRoomId]);
  console.log('Donte was here');

  const handleCreateRoom = () => addRoom();

  return (
    <HomePage>
      <header className="home-page-header">
        <img src={logo} className="home-page-logo" alt="logo" />
        <p>Create a room to get started</p>
        <Button color="primary" padding="2" onClick={handleCreateRoom}>
          {addRoomMutation.loading ? 'loading...' : 'create a room'}
        </Button>
      </header>
    </HomePage>
  );
};

Home.propTypes = {};

export default Home;
