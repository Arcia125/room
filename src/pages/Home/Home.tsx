import React, { ReactChild } from 'react';
import styled from 'styled-components';
import { useMutation, useQuery } from '@apollo/react-hooks';
import { RouteComponentProps } from 'react-router';
import { useHistory } from 'react-router-dom';

import { HomePage } from './styles';
import { Button } from '../../components/styles/Button';
import Navbar from '../../components/Navbar';
import GetStarted from '../../components/GetStarted';
import { GET_CURRENT_USER } from '../../graphql/getCurrentUser';
import { CREATE_USER } from '../../graphql/createUser';
import { SAVE_USER } from '../../graphql/saveUser';
import { useField } from '../../hooks/useField';
import { HelperText } from '../../components/styles/HelperText';
import { Link } from '../../components/styles/Link';

const Home: React.FunctionComponent<RouteComponentProps> = () => {
  const history = useHistory();
  const getCurrentUserQuery = useQuery(GET_CURRENT_USER);
  const currentUser =
    getCurrentUserQuery.data && getCurrentUserQuery.data.currentUser;

  if (getCurrentUserQuery.loading) return <HomePage>Loading...</HomePage>;
  if (getCurrentUserQuery.error) {
    return <HomePage>getCurrentUserQuery.error.message</HomePage>;
  }
  if (currentUser) history.push('/dashboard');

  return (
    <HomePage>
      <Navbar showLogin />
      <div className="homePage">
        <p className="homePage__invite">Pick a username to get started</p>
        <GetStarted />
        <HelperText>
          <span>Have an account?</span>
          <Link to="/login">Login</Link>
        </HelperText>
      </div>
    </HomePage>
  );
};

export default Home;
