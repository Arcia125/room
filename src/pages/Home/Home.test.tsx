import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { render, wait, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Home from '.';
import { Provider } from '../../theme';
import {
  currentUserMockResolvers,
  nullCurrentUserMockResolvers,
} from '../../../shared/queryMocks/currentUser';
import { mockRouteComponentProps } from '../../../shared/mocks/mockRouteComponentProps';

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(cleanup);

describe('Home', () => {
  it('renders without crashing when no user is logged in', async () => {
    const { container, asFragment } = render(
      <MockedProvider addTypename resolvers={nullCurrentUserMockResolvers}>
        <Provider>
          <Home {...mockRouteComponentProps} />
        </Provider>
      </MockedProvider>
    );

    await wait(() => expect(container).toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders without crashing when a user is logged in', async () => {
    const { container } = render(
      <MockedProvider addTypename resolvers={currentUserMockResolvers}>
        <Provider>
          <Home {...mockRouteComponentProps} />
        </Provider>
      </MockedProvider>
    );

    await wait(() => expect(container).toBeInTheDocument());
  });

  it('redirects logged in users', async () => {
    const { container } = render(
      <MockedProvider addTypename resolvers={currentUserMockResolvers}>
        <Provider>
          <Home {...mockRouteComponentProps} />
        </Provider>
      </MockedProvider>
    );

    await wait(() =>
      expect(mockRouteComponentProps.history.push).toBeCalledTimes(1)
    );
  });
});
