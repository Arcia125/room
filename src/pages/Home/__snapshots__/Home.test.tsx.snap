// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders without crashing when no user is logged in 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa ktYadu"
  >
    <nav
      class="sc-bwzfXH bPFVNE"
    >
      <a
        href="/"
      >
        <img
          alt="Room Logo"
          src="logoText.svg"
        />
      </a>
      <a
        class="sc-htpNat hGMLdq"
        color="#B3B2C0"
        href="/login"
      >
        LOGIN
      </a>
    </nav>
    <div
      class="homePage"
    >
      <p
        class="homePage__invite"
      >
        Pick a username to get started
      </p>
      <form
        class="sc-ifAKCX kCEAxF"
      >
        <div
          class="form"
        >
          <input
            class="form__input"
            placeholder="Enter a username"
            value=""
          />
          <button
            class="form__button sc-bxivhb esJnqo"
            color="primary"
            type="submit"
          >
            <i
              class="material-icons"
            >
              arrow_right_alt
            </i>
          </button>
        </div>
      </form>
      <p
        class="sc-EHOje hiuNXc"
      >
        <span>
          Have an account?
        </span>
        <a
          class="sc-htpNat eeHqNq"
          href="/login"
        >
          Login
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
